native/server/src/brongnal.rs

```diff
@@ -0,0 +1,201 @@
+            Status::unauthenticated("failed to validate one time prekey bundle signature")
+        })?;
+
+        self.storage.add_user(identity.clone(), ik, spk_proto)?;
+        // TODO(#25) - Figure about a better way to handle repeated registration.
+        self.storage.clear_one_time_keys(&identity)?;
+        self.storage.add_one_time_keys(&identity, pre_keys)?;
+
+        Ok(Response::new(RegisterPreKeyBundleResponse {}))
+    }
+
+    async fn request_pre_keys(
+        &self,
+        request: Request<RequestPreKeysRequest>,
+    ) -> Result<Response<PreKeyBundle>> {
+        let request = request.into_inner();
+        println!("Retrieving PreKeyBundle for \"{}\".", request.identity());
+
+        let (identity_key, signed_pre_key) = self.storage.get_current_keys(request.identity())?;
+        // TODO(#26) - Prevent one time key pop abuse.
+        let otk = self.storage.pop_one_time_key(request.identity())?;
+
+        let reply = PreKeyBundle {
+            identity_key: Some(identity_key.as_bytes().into()),
+            one_time_key: otk.map(|otk| otk.as_bytes().into()),
+            signed_pre_key: Some(signed_pre_key.into()),
+        };
+        Ok(Response::new(reply))
+    }
+
+    async fn send_message(
+        &self,
+        request: Request<SendMessageRequest>,
+    ) -> Result<Response<SendMessageResponse>> {
+        let request = request.into_inner();
+        println!(
+            "Sending a message to: \"{}\".",
+            request.recipient_identity()
+        );
+
+        let recipient_identity = request.recipient_identity.ok_or(Status::invalid_argument(
+            "request missing recipient_identity",
+        ))?;
+        let message_proto: proto::service::Message = request
+            .message
+            .ok_or(Status::invalid_argument("request missing message"))?
+            .into();
+        let _ = protocol::x3dh::Message::try_from(message_proto.clone())?;
+
+        let tx = self
+            .receivers
+            .lock()
+            .unwrap()
+            .get(&recipient_identity)
+            .map(|tx| tx.clone());
+        if let Some(tx) = tx {
+            if let Ok(()) = tx.send(Ok(message_proto.clone())).await {
+                return Ok(Response::new(SendMessageResponse {}));
+            }
+        }
+
+        self.storage
+            .add_message(&recipient_identity, message_proto)?;
+        Ok(Response::new(SendMessageResponse {}))
+    }
+
+    type RetrieveMessagesStream = ReceiverStream<Result<proto::service::Message>>;
+    async fn retrieve_messages(
+        &self,
+        request: Request<RetrieveMessagesRequest>,
+    ) -> Result<Response<Self::RetrieveMessagesStream>> {
+        let request = request.into_inner();
+        println!("Retrieving \"{}\"'s messages.", request.identity());
+
+        let identity = request
+            .identity
+            .ok_or(Status::invalid_argument("request missing identity"))?;
+        let (tx, rx) = mpsc::channel(100);
```

Comment:
magic number, what is 100?

I see the value here https://docs.rs/tokio/latest/tokio/sync/mpsc/fn.channel.html, but does buffer size matter here?

Reply:
The async runtime blocks while waiting for the consumer to retrieve messages if the queue is full. 100 is reasonable for now